### ISSUE TITLE
Proposal: a sidebar tab where we can save and restore stories

### ISSUE THREAD

**Issue by BurningTreeC** [Jermolene/TiddlyWiki5#4319]
The proposal would be a sidebar tab called `Stories` where we can save a current StoryList and restore saved StoryLists

**Comment by Jermolene** [comment]
Hi @BurningTreeC I'm definitely interested in adding such story handling, but I don't think it should be a top level sidebar tab. It's the kind of thing that wouldn't make much sense to a first time user; it's only once the concept of the story river has bedded down that a user would understand the need. So I'd favour putting it under the "More" sidebar tab.

**Comment by Jermolene** [comment]
It also occurs to me that there is a relationship with the permaview functionality, which is also presented as a way to share and save story river content.

**Comment by BurningTreeC** [comment]
Hi @Jermolene, the `More` Sidebar tab makes sense to me

**Comment by twMat** [comment]
>but I don't think it should be a top level sidebar tab. It's the kind of thing that wouldn't make much sense to a first time user

This thread inspired me to post #4323

**Comment by AnthonyMuscio** [comment]
I posted in Proposal: High level UI for add/remove sidebar tabs #4323 but was keen to get some story saving as well.

Some thoughts

I was thinking what If I could save the current story, or as Jeremy suggested the current permaview against the current tiddler say in $:/storylist/currenttiddler and on opening such a tiddler, a button would appear to open that tiddlers saved story and either replace/add to the current story.

Because I wanted to do this for the current Project tiddler I thought about adding a filter as well sto save to the project story the current tiddler and any open tiddler with project=current project.

A little more thought may come up with a more broadly applicable solution and Proposal: High level UI for add/remove sidebar tabs #4323 can determine where it is displayed.

**Comment by pmario** [comment]
I want to be able to define a story list similar to the DefaultTiddlers definition. That means, it should be possible to mix tiddler names with filters. eg: 

```
Learning [tag[HelloThere]] 
Community
```

If you copy paste the above code into the DefaultTiddlers section in the ControlPanel and hit the "home-button" in the sidebar it works. 

There is a `tm-home` event, that creates this functionality. The tm-home event doesn't allow any parameters. 

So if we create a `tm-story` message, that allows us to specify 

 - the tiddler, that contains the filter
 - the story-list tiddler, that should be modified
 - the history tiddler that should be modified

I think we are good to go. .... I think, I'll test this with a plugin.

**Comment by pmario** [comment]
Having a closer look, I found out that tiddlers tagged `$:/tags/Filter` technically are already stories. 

There are some use-case differences, but we should keep this in mind, before we define new system tags. ... They may be not needed. ... 

More thinking is needed.

**Comment by pmario** [comment]
This needs to work too: 

```
[[Discover TiddlyWiki]] [[A Gentle Guide to TiddlyWiki]] +[sortby{HelloThere!!list}]
```

So a story, that contains an unordered subset of the "full story" should be able to use the sort order of the full story.

**Comment by Jermolene** [comment]
Hi @pmario I think you're pointing out that we need to be able to define a story by specifying a filter. I agree.

**Comment by AnthonyMuscio** [comment]
I think a more generic approach could help.

As the current story is stored in the list field of $:/StoryList tiddler cloning this tiddler or field captures the current story. 

A macro such as 
\define current-story() {{$:/StoryList!!list}}

current-story Should always return the current story. 
Lists are also a form of filter, and can be filtered them self however rather than filter the titles in a list,

What if we had a simple mechanisium to store a variable such as current-story for later use. Ie a saved list. Perhaps under $:/savedlists/ then the ability to use any saved list, optionaly applying a filter based on the tiddlers named in the list. Ie not on the titles but the tiddlers represented by those titles.

Then a way to set the current story to the result, and opening them in the story. We would need an option to close all first or not.

The ability to chain stories with next and previouse, options for single tiddler views and more becomes possible.

With this building in a stories side bar becomes trivial.

**Comment by Jermolene** [comment]
Hi @AnthonyMuscio your enthusiasm and commitment really is inspiring, but I fear that sometimes your posts don't always get across what you're trying to say as clearly as they might.

Here, you've got some typos/formatting issues which tend to indicate a failure to read things through before you post. Remember that if you post here many people are going to get the email notification, and give their attention to what you've written (not all will click through to see the original post which is why it's so important to get things right first time as much as possible). If we're going to demand attention then we owe our collaborators the courtesy of taking the trouble to communicate as clearly as possible, and to carefully check what we write before we post (this post has taken me 20 minutes, for example).

I'm going to give more direct feedback than I would usually, I hope it's helpful. I appreciate your contributions very much and hesitate to appear to criticise anyone, please be assured that this is intended constructively.

> I think a more generic approach could help.

You may not be aware, but this reads like the purpose of your post is to disagree with something that a previous poster has said. If you open in that way then your correspondents will be reading looking out for the thing that you think that they've got wrong. If you don't mean to signal disagreement then it might be better to say something more like "Great idea, perhaps we could build on it to...".

> What if we had a simple mechanisium to store a variable such as current-story for later use. Ie a saved list. Perhaps under $:/savedlists/ then the ability to use any saved list, optionaly

I'm struggling to parse exactly what you mean from this point on. You propose a mechanism to store a variable for later use, but there's no clear meaning to "storing variables", perhaps you mean take the value from a variable and store it somewhere?

I get the idea that you're thinking of something very specific (e.g. some special behaviour for tiddlers with the title prefix you mentioned). If you gave an example of such a tiddler things would be much clearer.

> applying a filter based on the tiddlers named in the list. Ie not on the titles but the tiddlers represented by those titles.

This distinction is surely at the level of filter operators? Filters always return a list of titles, and many filter operators work directly on those titles but some filter operators refer to the tiddler store to access the tiddlers represented by those titles.

> The ability to chain stories with next and previouse, options for single tiddler views and more becomes possible.

I think you'd need to expand on those, particularly the single tiddler view idea.

**Comment by AnthonyMuscio** [comment]
Jeremy I appreciate constructive critisisium and suggestions. In my defence I responded after rereading the full thread but the brevity and errors are most likely due to replying from my mobile. As a result I will avoid compounding the problem and reply when back on my computer.

**Comment by pmario** [comment]
@Jermolene ... In my proposal and proof of concept TW, a story is technically the same as a AdvancedSearch filter string

Those "filter strings" also work, if you add them without any editing to the $:/DefaultTiddlers as a "startup" configuration. 

The Filter-tab already contains 14 configurations in the filter dropdown. Those elements are tagged: `$:/tags/Filter`

My [bundler-plugin](https://wikilabs.github.io/editions/bundler/#%24%3A%2FAdvancedSearch) also uses this tag for "filtered-bundles". ... 

**The problem is**, that the resulting dropdown list is complicated to read. ... 

So should we use a new system tag eg: `$:/tags/Story` or create a different list view? ... I'll try a "tree-view" in the POC. ... link will be posted soon.

**Comment by pmario** [comment]
first screenshot: 

![story-concept-01](https://user-images.githubusercontent.com/374655/67389671-5cf7f200-f59b-11e9-856e-6bf195b24ccc.gif)

So if users do proper grouping, it should be possible to manage the list. Story tiddlers use a namespace: `Story/group/subgroup/title with spaces`

**Comment by pmario** [comment]
Starting a TW with a story is as simple as: Set `$:/DefaultTiddlers` to eg: `[subfilter{Story/aa/test 20}]`

**Comment by BurningTreeC** [comment]
@pmario, this looks interesting! in your example, is it possible to save the current storylist as such list?

**Comment by Jermolene** [comment]
Hi @pmario the primary consequence of using filters to represent trails is that users cannot drag and drop to add/reorder entries within a filter, only to add/reorder entries to a list. So I think canned filters and trails are different things, both useful in different situations.

**Comment by pmario** [comment]
@Jermolene ... For me it's OK if a "trail" is a different thing as a "story". 

If the term "trail" defines, that is has to have a list-field or a list-like-field I'm OK with that. list-like-fields are a subset of the TW filter notation. So trails are compatible with stories.

eg: A list field, that contains: `a b c` is a separated list. So it can be **used as a list** or **as a filter**.

In my example, I wanted to show, that for me a "story" needs to be much more dynamic. ... 

Since filters and lists are different things, that means, that the **main field** for a story can't be the `list` field, but it's perfectly fine to be the **default field** for many users. 

For advanced users it may be a "filter-like" field

**Comment by pmario** [comment]
I think a Story tiddler could have the following fields, which would allow a consistent behaviour and it would be compatible with existing TW UI tools

 - title: <anyName> ... in my example they are "namespaced", but that's just a convention
 - filter: ... defines a Story
 - list: ... defines the "default" Trail
 - text: Some basic info about "the thing" (trail or story)

 - description: 
      - optional ... for trail or story
      - mandatory ... for the AdvancedSearch:Filter tab
 - tags: 
     - optional ... for story ... 
     - optional or mandatory ... I have no idea about "trails" atm
     - mandatory ... $:/tags/Filter for AdvancedSearch:Filter tab
 
  - caption
     - optional for "lists" and "TOCs" ... tree macros don't support it YET

**Comment by AnthonyMuscio** [comment]
As raised elsewhere I would like to see the following available, needs minor change to the standard search and would support story tiddlers.

- system-caption: optional will be found in the "standard search" results even for system tiddlers, system-caption will be displayed in search results rather than the system tiddler title
- search-keywords: optional will be found in the standard search results even for system tiddlers

So if someone has a $:/storytiddler/name and they search for the story "name" eg project it will be found in the standard search if system-caption has "Project1 story name", or the search keywords contain "projectname story work in progress", at which point it will also list the tiddler as "Project1 story"

Regards
Tony

**Comment by BurningTreeC** [comment]
Hi @pmario, I'm curious about your implementation how you load/unload stories using the list and filter fields

**Comment by pmario** [comment]
That's simple. The tw-core provides functions for that. I'm using a new `tm-open-story` message. 

The advantage will be, that everyone can make new UI elements. So I only need to add "just enough" UI elements to get it going.

**Comment by pmario** [comment]
If everything works as expected, it should be possibile to open "bundler" configurations as stories.

**Comment by pmario** [comment]
There is 1 more point, that was partially mentioned at the very beginning by Jeremy. **permaview** and / or the update of the browser URL and the browser history.

TW has several settings, that influence the browser history. See ControlPanel: Settings

### Navigation Address Bar

Behaviour of the browser address bar when navigating to a tiddler:

 - Include the target tiddler and the current story sequence  <- default
 - Include the target tiddler
 - Do not update the address bar

### Navigation History

Update browser history when navigating to a tiddler:

 - Update history
 - Do not update history <- default

----------

So from my point of view, if "Navigation History" is set to "Update history" ... opening a "saved story" should also update the browser history accordingly. 

At the moment it's not clear to me, what "accordingly" should do. 

May be we need a new setting like: - Update browser history, only if new story has been loaded. 

This would allow us to "jump" forward and backward on stories ... **using the browser** forward and backward buttons.

**Comment by AnthonyMuscio** [comment]
Folks,

I am continuing to look into stories. I thought I would share what I am working on. Basically it provides a logical method to name a story according to its place "in context".

This allows the user to save multiple stories according to the work they are doing, and be reminded when a saved story exists. I hope I could leverage @pmario Story tiddler standard, and the bundler plugin. Once could open a saved story, open some more tiddlers then save it as a bundle.

I have a wiki in which most tiddlers have one or more of three fields domain, project and client
I have a variable that allows me to create new tiddlers "in context" I use this in new tiddler buttons.
`\define context() domain="{{!!domain}}" project="{{!!project}}" client="{{!!client}}"`

So in an attempt to save stories in a logical manner, I thought lets create a "context" tiddler for each context and use a tiddler named as follows; but save a story in its list field.
`\define context-tiddler() <$wikify name=result text="""$:/contextTiddlers/{{!!domain}}/{{!!project}}/{{!!client}}"""><<result>></$wikify>`

- So now I am building a button that on clicking will copy the tiddlers named in $:/StoryList!!list to `<<context-tiddler>>!!list ` only if they share the same context. Ie any open tiddler, that has the same context will be saved in a tiddler by the name of the context.
- I will also create a button called close story/context, that does the same as above but closes all tiddlers in the same context story tiddlers as well.
- One final button that can appear on any tiddler, "with a context" says open current context story (if it exists).

This method allows one to forget the story name altogether and access it by opening a tiddler in the context that relates to the saved story.

Now over in the sidebar **as in the original thread**, I can provide options to open all tiddlers in saved stories by selected context, or others such as all tiddlers in stories for a domain and any domain story (regardless or project and client), then also for project and client. Perhaps these could also be independent story tiddlers (with filters).

This idea can be incorporated into what is being discussed in this thread. Including providing alternative filtered stories, so for example open all unclosed tasks in the in context story. This filtered story can appear "in context". 

I hope I have being sufficiently eloquent, feel free to ask for clarification.

**Comment by AnthonyMuscio** [comment]
Oh, and rather than my suggested context one could have a storyname field used as the context, containing one or more stories to which a tiddler belongs.

**Comment by pmario** [comment]
I'll post a link to a working prototype soon.

**Comment by pmario** [comment]
>I'll post a link to a working prototype soon.

hmmm, a bit late but finally: 

 - Video: https://www.youtube.com/playlist?list=PLuiC_HFhI4OwuaGpNyloRZ3zdJWmrFO_J
 - DEMO edition: https://wikilabs.github.io/editions/wltm-open-story
 - Voting: https://github.com/wikilabs/plugins/issues/43

**Comment by twMat** [comment]
@pmario - great stuff!!! Before voting, I think there needs to be some discussion about it. Overall I do think some kind of story feature makes a lot of sense in the core but exactly how it is manifested is not as obvious. Where is an appropriate place to discuss some of the intricacies in your creation?

**Comment by pmario** [comment]
At the github issues and: https://groups.google.com/forum/#!topic/tiddlywiki/z7938E0bi2k
